src/Voron/Impl/Paging/CryptoPager.cs

```diff
@@ -114,12 +113,28 @@ private static int GetNumberOfPages(PageHeader* header)
             return checked((overflowSize / Constants.Storage.PageSize) + (overflowSize % Constants.Storage.PageSize == 0 ? 0 : 1));
         }
 
-        public override byte* AcquirePagePointerForNewPage(IPagerLevelTransactionState tx, long pageNumber, int numberOfPages, PagerState pagerState = null)
+        public override byte* AcquirePagePointerForNewPage(IPagerLevelTransactionState tx, long pageNumber, int numberOfPages, PagerState pagerState = null, bool handleOverwrite = false)
         {
-            // New page -> no need to read page, just allocate a new buffer
             var state = GetTransactionState(tx);
             var size = numberOfPages * Constants.Storage.PageSize;
-            var buffer = GetBufferAndAddToTxState(pageNumber, state, size);
+
+            if (handleOverwrite && state.LoadedBuffers.TryGetValue(pageNumber, out var buffer))
+            {
+                if (size == buffer.Size)
+                {
+                    // precaution
+                    Memory.Set(buffer.Pointer, 0, size);
+                    Memory.Set(buffer.Hash, 0, EncryptionBuffer.HashSizeInt);
```

Comment:
Don't know if that's necessary

Reply:
It _probably_ isn't, but it doesn't hurt, and I can see some scenarios where it would be useful

Reply:
Just noticed we call `Sodium.sodium_memzero(ptr, (UIntPtr)size);` when returning the buffer to the pool:
https://github.com/ravendb/ravendb/blob/v4.0/src/Voron/Impl/EncryptionBuffersPool.cs#L68

So I'll call it here as well, instead of `Memory.Set` for consistency.

Reply:
Yes, that is better